src/transformStateWithClones.js

```diff
@@ -6,8 +6,43 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  for (const action of actions) {
+    const { type = null, extraData = {}, keysToRemove = [] } = action;
+
+    let newObj;
+
+    if (states.length === 0) {
+      newObj = { ...state };
+    } else {
+      newObj = { ...states[states.length - 1] };
+    }
+
+    switch (type) {
+      case 'addProperties':
+        newObj = Object.assign(newObj, extraData);
+        break;
+      case 'removeProperties':
+        keysToRemove.forEach((element) => {
+          if (newObj[element]) {
+            delete newObj[element];
+          }
+        });
+        break;
+      case 'clear':
+        Object.keys(newObj).forEach((element) => {
+          delete newObj[element];
+        });
+        break;
+    }
```

Comment:
The switch statement should include a `default` case to handle any unexpected action types. This is important for error handling and to ensure that the function behaves predictably even with unexpected input.